src/mapblock_mesh.cpp

```diff
@@ -194,37 +195,74 @@ u16 getFaceLight(MapNode n, MapNode n2, v3s16 face_dir, INodeDefManager *ndef)
 	Calculate smooth lighting at the XYZ- corner of p.
 	Both light banks
 */
-static u16 getSmoothLightCombined(const v3s16 &p, MeshMakeData *data)
+static u16 getSmoothLightCombined(const v3s16 &p,
+	const std::array<v3s16,8> &dirs, MeshMakeData *data, bool node_solid)
 {
-	static const v3s16 dirs8[8] = {
-		v3s16(0,0,0),
-		v3s16(0,0,1),
-		v3s16(0,1,0),
-		v3s16(0,1,1),
-		v3s16(1,0,0),
-		v3s16(1,1,0),
-		v3s16(1,0,1),
-		v3s16(1,1,1),
-	};
-
 	INodeDefManager *ndef = data->m_client->ndef();
 
 	u16 ambient_occlusion = 0;
 	u16 light_count = 0;
 	u8 light_source_max = 0;
 	u16 light_day = 0;
 	u16 light_night = 0;
+	bool corner_obstructed = true;
+	bool index4_obstructed = true;
+	bool index5_obstructed = true;
+	bool index6_obstructed = true;
+	bool index7_obstructed = true;
+
+	static thread_local const bool edge_obstruction =
+		g_settings->getBool("smooth_lighting_edge_obstruction");
+
+	for (int i = 0; i < 8; ++i) {
+		if (edge_obstruction && !node_solid && ((i == 4 && index4_obstructed) ||
+				(i == 5 && index5_obstructed) ||
+				(i == 6 && index6_obstructed) ||
+				(i == 7 && index7_obstructed))) {
+			ambient_occlusion++;
+			continue;
+		}
 
-	for (const v3s16 &dir : dirs8) {
-		MapNode n = data->m_vmanip.getNodeNoExNoEmerge(p - dir);
+		MapNode n = data->m_vmanip.getNodeNoExNoEmerge(p + dirs[i]);
 
 		// if it's CONTENT_IGNORE we can't do any light calculations
 		if (n.getContent() == CONTENT_IGNORE)
 			continue;
 
 		const ContentFeatures &f = ndef->get(n);
+		if (edge_obstruction) {
+			if (node_solid) {
+				if ((i == 4 && corner_obstructed)) {
+					ambient_occlusion++;
+					continue;
+				}
+				// Only the 4 nodes in front of the face (and
+				// the corner's node) can contribute light
+				if (i > 4) {
+					break;
+				} else if (i < 2 && f.param_type == CPT_LIGHT){
+					corner_obstructed = false;
+				}
```

Comment:
braces can be removed for this if construct

Reply:
Ok